src/gsi/read_gnssrspd.f90

```diff
@@ -0,0 +1,514 @@
+           if (regional) then
+              call tll2xy(dlon_earth,dlat_earth,dlon,dlat,outside)
+              if (diagnostic_reg) then
+                 call txy2ll(dlon,dlat,rlon00,rlat00)
+                 ntest      = ntest+1
+                 cdist      = sin(dlat_earth)*sin(rlat00)+cos(dlat_earth)*cos(rlat00)* &
+                             (sin(dlon_earth)*sin(rlon00)+cos(dlon_earth)*cos(rlon00))
+                 cdist      = max(-one,min(cdist,one))
+                 disterr    = acos(cdist)*rad2deg
+                 disterrmax = max(disterrmax,disterr)
+              end if
+              if(outside) cycle loop_readsb2
+           else
+              dlon = dlon_earth
+              dlat = dlat_earth
+              call grdcrd1(dlat,rlats,nlat,1)
+              call grdcrd1(dlon,rlons,nlon,1)
+           endif
+
+
+!          Read surface wind speed [m/s] from GNSSRSPD 
+           if (lspdob) then
+              !usage=r100 
+              !usage = zero  ! will be considered for assimilation
+
+              ! Get Wind Speed observations from bufr file
+              call ufbint(lunin,gnssrw,1,1,nlv,wndstr)
+              spdob = gnssrw(1,1) ! surface wind speed 
+
+              ! Don't permit observations with ws <= 1 m/s
+              if (spdob <= 1.0) cycle loop_readsb2
+              if (spdob >= missing) cycle loop_readsb2
+
+              ! Get observation error from bufr file
+              call ufbint(lunin,gnssrw,1,1,nlv,oestr)
+              obserr = max(gnssrw(1,1),1.5) ! surface wind speed error
+           endif
+
+
+           if ( .not. twodvar_regional) then
+              call deter_sfc_type(dlat_earth,dlon_earth,t4dv,isflg,tsavg)
+           endif
+
+           ! Get information from surface file necessary for conventional data
+           call deter_sfc2(dlat_earth,dlon_earth,t4dv,idomsfc,tsavg,ff10,sfcr,zz)                                                                      
+           ! Process data passed quality control 
+           igood = igood + 1
+           ndata = ndata + 1
+           nodata = nodata + 1
+           iout = ndata
+
+!-------------------------------------------------------------------------------------------------          
+
+           ! Winds --- surface wind speed 
+           if (lspdob) then
+              woe = obserr
+              !if (inflate_error) woe = woe*r3_0
+              !if (inflate_error) woe = woe*r1_2
+              !if (qcm > lim_qm ) woe = woe*1.0e6_r_kind
+              cdata_all( 1,iout)=woe                    ! wind error
+              cdata_all( 2,iout)=dlon                   ! grid relative longitude             
+              cdata_all( 3,iout)=dlat                   ! grid relative latitude                  
+              cdata_all( 4,iout)=dlnpsob                ! ln(surface pressure in cb)
+              cdata_all( 5,iout)=spdob*sqrt(two)*half   ! u obs
+              cdata_all( 6,iout)=spdob*sqrt(two)*half   ! v obs
```

Comment:
@KariA-Spire 
The observation is `wind speed` (spdob).
The u-wind (uob) and v-wind (vob) components are derived by assuming uob = vob.
And, the uob and vob will always be positive.  
I am curious: what is the scientific basis for this assumption?

Reply:
As a first approximation for assimilating GNSS-R wind speed observations, some members of the CYGNSS science team suggested using existing GSI code for reading ocean surface wind speed from a Step Frequency Microwave Radiometer (SFMR).  There are multiple well-known meteorological scenarios where the zonal and meridional wind components are equal in magnitude at the ocean surface. For example, in the inter-tropical convergence zone trade winds from the subtropics converge with winds originating from the opposite direction, the resulting wind field can have a 45-degree angle with respect to the u and v components, making them equal in magnitude. Similarly, during tropical cycles, storm displacement can lead to diagonal wind fields relative to the zonal and meridional wind components. Analogous to SFMR wind speed observations, GNSS-R observations can benefit tropical weather observation and prediction. Moreover, this calculation was chosen while the GNSS-R community continues to develop and test a wind direction retrieval algorithm or a machine-learned estimation to allow for a more precise remotely sensed measurement of the zonal and meridional wind components.

Reply:
@KariA-Spire Thanks for your explanation and the slides.  
I understand that the assimilation of GNSS-R wind speed is following the footstep of FSMR retrieved wind speed, which also assumes uob and vob.  I believe that must be QC associated with this assumption for FSMR in GSI.

Thanks for your clarification.  It is good to hear that the wind direction retrieval is in progress.  
I am good with your changes.

Reply:
@emilyhcliu, thank you very much for reviewing this PR and for your insightful comments.

Reply:
@emilyhcliu , if this conversation is resolved, please mark it as so.  We can not process GSI PRs with unresolved coversations.